### DISCUSSION TITLE
New args: sentence, standard, max_line_width, max_line_count, max_comma, max_gap.

### DISCUSSION THREAD

**Discussion by Purfview** [Purfview/whisper-standalone-win#125] [Announcements]
Post feedback about these here.

New parameters :
`--sentence`
`--standard`
`--max_comma`
`--max_gap`
`--max_line_width`
`--max_line_count`

Implemented since r160.8.

r160.9 UPDATE:

Bugfix: The final word could disappear in some combination of the new parameters from r160.8.
New parameter : `--standard_asia`

r160.10 UPDATE:

Bugfix: The final line disappeared if the final word wasn't punctuated when using  `--sentence`.

**Comment by BenJamesAndo** [comment]
Thanks for adding these! I'm now able to get very close to where I'd like to get.
Firstly a bit of feedback on the help file. It says under `--sentence`
... Be default meant to output whole sentence per line for better translations, but not limited to. (default: False)

I'm not quite sure what that means. Is it supposed to say something like: The default behaviour is to add a whole sentence per line for better translations, but not be strictly limited by it?

If I run a command with  `--max_line_width 37 --max_line_count 2` then I get

Error: Original functionality of 'max_line_width' and 'max_line_count' parameters is borked.
       Only the alternative funtionality is allowed with `--sentence`.
       
Is this saying that `--sentence` is mandatory when using `--max_line_width` & `--max_line_count`? Because it doesn't say that it's required in the help file.

I've tested this out on a few 28 minute videos. While it technically works flawlessly, it does lead to another problem. Sometimes I have a line with only one word in it.
![image](https://github.com/Purfview/whisper-standalone-win/assets/44917135/21097839-0a29-4177-92c7-0fc30e37e732)

![image](https://github.com/Purfview/whisper-standalone-win/assets/44917135/bbd46b44-31b1-4da6-9b67-1e90a384a838)

I wonder if it's worth having a new parameter that specifies the minimum amount of characters on a line. e.g. `--min_line_width 8`. It could get quite complicated as it may not always be able to fulfil the requirements. But in the above examples, it could certainly move a few words from the first section into the section with only one word.

**Reply by Purfview** [reply]
> Is this saying that --sentence is mandatory when using --max_line_width & --max_line_count?

Yes.

> I wonder if it's worth having a new parameter... "--min_line_width"

No, as it doesn't make sense.

**Reply by BenJamesAndo** [reply]
So do you think for my situation I'm best to just stick with `--one_word` and then use Subtitle Edit to merge the lines together? Or is there some other way to avoid having only one word in a line, whilst still adhering to a certain max line width and max line count?

**Reply by Purfview** [reply]
You say it like SE has `min_line_width` option or like you wouldn't get those one words there too.

`min_line_width` would break many other things. Tell me this magic what to do if `min_line_width` would be met in your examples.

**Reply by BenJamesAndo** [reply]
Sorry I misread your response https://github.com/SubtitleEdit/subtitleedit/issues/7655#issuecomment-1832935878
I thought you were implying that these new options would help with the one word on a line problem.
You are right, using SE via CLI also produces the same result.

Intererestingly though, generating captions with Faster-Whisper without any arguments directly in the SE GUI, it actually does somehow adhere to the single line max. length. that's set in the general settings.
It's because of having Auto adjust timings option enabled, which doesn't seem accessible via the CLI. I haven't done extensive testing, but it does seem to do quite a good job. I'm just trying to get the same results through only the CLI.

**Reply by Purfview** [reply]
> I thought you were implying that these new options would help with the one word on a line problem.

That thread is about merging `--one_word`, which is the option from Standalone Faster-Whisper, not about "one word problem".

> Intererestingly though, generating captions with Faster-Whisper without any arguments directly in the SE GUI...

I think that's because of "post-process" option is selected. I personally deselect all there as it does something weird. I think those were meant for some poor whisper implementations, I recommend to disable them all with Standalone  Faster-Whisper.

**Reply by oep42** [reply]
@BenJamesAndo I suggested a tool to add to Subtitle Edit, to automatically solve the "short line" or "one word on a line" problem after subtitles have been created with a Whisper engine. I hope that a tool like this will be added to Subtitle Edit. See [here](https://github.com/SubtitleEdit/subtitleedit/issues/8714#issue-2454128744).

**Comment by ext666** [comment]
Looks great, thanks.

Would it be possible to make a command that would limit the maximum length of the subtitle in seconds? Something like `max_time 6` (for 6 seconds).

**Reply by Purfview** [reply]
> Would it be possible to make a command that would limit the maximum length of the subtitle in seconds? Something like max_time 6 (for 6 seconds).

Possible, what settings do you use and could you provide json file for example?

**Comment by despairTK** [comment]
Hi! After I tried the new parameter --sentence, I found it very useful. But I also have some doubts.

For example:

337
00:34:27,020 --> 00:34:40,420
What I want you to focus on and what I'm trying to focus on with this particular study is a breakdown of the lights and the darks and an indication of these things with very, very simple shapes because I'm going to start moving forward on the painting now,

338
00:34:40,480 --> 00:34:43,980
and I want you to look at how I proceed.

Before adjustment:
[34:27.020 --> 34:31.640] What I want you to focus on and what I'm trying to focus on with this particular study is
Processing segment at 31:01.260
[34:32.040 --> 34:37.180] a breakdown of the lights and the darks and an indication of these things with very, very
[34:37.180 --> 34:40.940] simple shapes because I'm going to start moving forward on the painting now, and I want you
[34:40.940 --> 34:43.980] to look at how I proceed.

This is also true in another place:

340
00:34:51,120 --> 00:35:12,220
I'm going to kind of average the skin tone of his face and start there, and I want you to see how quickly, because I have these shadow shapes established and I have the drawing in place done by putting these value shapes in their place , I can actually focus wholly now on the transition areas,

341
00:35:12,320 --> 00:35:16,080
and also I can focus on putting really expressive brushwork in here.

Before adjustment:
[34:51.120 --> 34:57.340] I'm going to kind of average the skin tone of his face and start there, and I want you
Processing segment at 31:26.960
[34:57.340 --> 35:02.720] to see how quickly, because I have these shadow shapes established and I have the
[35:02.720 --> 35:08.920] drawing in place done by putting these value shapes in their place, I can actually focus
[35:08.920 --> 35:15.680] wholly now on the transition areas, and also I can focus on putting really expressive brushwork
[35:15.680 --> 35:16.080] in here.

The above two examples do not form a complete sentence, but are separated by commas into two lines of subtitles. Although it does not affect translation, sentences with more words than them are well segmented. Don't know if this is normal.

**Reply by Purfview** [reply]
Read about these thresholds:
--max_comma
--max_gap

In your case you hit 250 default of --max_comma, set it to same as --max_line_width to deactivate it.

EDIT:
Btw, --standard probably is not good for the languages with hieroglyphs, I can add other presets if you need.

**Reply by despairTK** [reply]
Hieroglyphs? Does it mean emojis or some Asian language? Such as Chinese, Japanese, Korean and so on?

In addition, I have an idea. A few days ago, I tested a method that can more accurately transcribe Korean and Japanese audio. Just specify --language English to transcribe Korean and Japanese audio directly when specifying the language. The subtitles transcribed in this way will largely avoid the incorrect selection of some polyphonic words, and the accuracy is much higher than that of --language JP/KR. But doing so requires some prerequisites.

1. You need a Japanese subtitle that has been transcribed, then select all subtitle lines in Subtitle Edit to transcribe line by line, and set the --language to English. (This method is best used on the original Japanese subtitle file. Each sentence is a longer sentence. If the sentence is too short, it may be transcribed into transliteration or Japanese.)

2. Directly select the entire audio as a line of subtitles in Subtitle Edit, and then transcribe this line of subtitles. This will get 100% English subtitles, but there is no timeline. If --standard can segment the result, it will be perfect. of.(I also tested Spanish, and this method still provided highly accurate results.)

**Reply by Purfview** [reply]
> Such as Chinese, Japanese, Korean and so on?

Yes.
Some places say to use 16 some 23.

> In addition, I have an idea. 

I can't make sense of it. Anyway, it doesn't belong in this thread.

**Reply by despairTK** [reply]
What do 16 and 23 stand for? Is --max_gap?

**Reply by Purfview** [reply]
Maximum characters per line of course. Why would you think that it means anything else?
I was talking about `--standard` preset.

**Reply by despairTK** [reply]
terribly sorry. . . I misread the word in your reply.

Regarding --standard , I tested it and I don’t think this function will be used. It breaks up otherwise complete sentences and threatens to make Arabic numerals disappear from the subtitles.

The following is an example
Transcription settings: --model large-v2 --beam_size 10 --best_of 10

Before processing:
![QQ截图20231201221448](https://github.com/Purfview/whisper-standalone-win/assets/111871110/2ec51f3a-1965-4292-b18d-91814e27ca06)
--sentence ture:
![QQ截图20231201221659](https://github.com/Purfview/whisper-standalone-win/assets/111871110/85235e60-28c4-480a-992b-457dfe05d151)
--standard ture:
![QQ截图20231201221709](https://github.com/Purfview/whisper-standalone-win/assets/111871110/b043a551-26a6-4e2a-a9d7-86846c1c0760)
The number [101] at the end of the last sentence disappeared.

**Reply by Purfview** [reply]
> Regarding --standard , I tested it and I don’t think this function will be used. It breaks up otherwise complete sentences

That's exactly what it should do, did you read --help?

> disappear from the subtitles.

That must be a bug. Can you share the `.json` file from those subs?

**Reply by Purfview** [reply]
"Wonderful Stylized Backgrounds and Planar" is 41 chars, + "101" would be over 42 chars.
Isn't "101" in the next line?
Your screenshot doesn't show what is after 32 line, stop confusing me.  😆

**Reply by despairTK** [reply]
it's here
[01.zip](https://github.com/Purfview/whisper-standalone-win/files/13527894/01.zip)

To be honest, I will read the --help information. However, since machine translation cannot accurately translate some --help information, I can only use audio testing to understand what each parameter does. . . .

Add a screenshot and audio, you can also test it.
[Audio.zip](https://github.com/Purfview/whisper-standalone-win/files/13527965/Audio.zip)
![QQ截图20231201225027](https://github.com/Purfview/whisper-standalone-win/assets/111871110/7180f34c-9a55-4389-be2b-6cbcf19f4de8)

**Reply by Purfview** [reply]
Yeah it's the bug, can happen only with last word in the whole file.

**Reply by Purfview** [reply]
Test if it's fixed in r160.9

**Reply by despairTK** [reply]
After testing, this problem was indeed fixed.

**Comment by Purfview** [comment]
I think new parameter for comma should be introduced, for example if comma is near at the line's end then it should break the new line, I meant don't add one or two short words after it.

**Reply by ClaireCJS** [reply]
> I think new parameter for comma should be introduced, for example if comma is near at the line's end then it should break the new line, I meant don't add one or two short words after it.

I've gone as far as to have my downloaded-lyrics-postprocessor add a comma at the end of every line just to try to somehow make it happen, so yes, I agree with this idea!!! 🎉

**Reply by Purfview** [reply]
That is already implemented: `--max_comma_cent`

**Comment by 1001ruchka** [comment]
Tested the operation with the - "--sentence" parameter.
When recognizing speech for the same video.  When using the parameter "--sentence", the last sentence in the video was not recognized.
Without the parameter "--sentence" it was recognized without any problems..

**Reply by Purfview** [reply]
"Recognized" and not present in srt can be two different things.
What version do you use and is that last sentence missing in txt file too?

**Reply by 1001ruchka** [reply]
Standalone Faster-Whisper r160.9 running on: CUDA

At the same time, in the log file, when running without the parameter, the missing phrase is also missing - but in the subtitles in SE it is there.

**Reply by 1001ruchka** [reply]
Tested it on another video - there the last sentence is not missing when using the parameter "--sentence".
I suspect the problem is that before the missing sentence in the previous video, there was a huge sentence without punctuation marks.

**Reply by Purfview** [reply]
Read the question again https://github.com/Purfview/whisper-standalone-win/discussions/125#discussioncomment-7741225

**Reply by 1001ruchka** [reply]
I didn't understand the question about which txt file you are talking about, I use Faster-Whisper in SE.
I've posted the audio of the problem here:

SE: 4.0.2.54 - Microsoft Windows NT 10.0.19045.0 - 64-bit
Message: Calling whisper (Purfview's Faster-Whisper) with : C:\Users\alexa\AppData\Roaming\Subtitle Edit\Whisper\Purfview-Whisper-Faster\whisper-faster.exe --language en --model "medium.en" --sentence "C:\Users\alexa\AppData\Local\Temp\48e9bfbf-d8b6-4af0-b05f-53b4f586164f.wav"
Standalone Faster-Whisper r160.9 running on: CUDA

**Reply by Purfview** [reply]
> I didn't understand the question about which txt file you are talking about, I use Faster-Whisper in SE.

One of the outputs formats. With `-f=all` you can output them all. Share srt and json files.
I'm not interested in what you see in SE nor in audio file.

**Reply by 1001ruchka** [reply]
[Purfview-Whisper-Faster.zip](https://github.com/Purfview/whisper-standalone-win/files/13537795/Purfview-Whisper-Faster.zip)
archive with all files

**Reply by Purfview** [reply]
Thx, now I see that it's some bug. I'll look at the code.

**Reply by Purfview** [reply]
Should be fixed in `r160.10`, hopefully no more bugs. 🪲

Good way to check:
With SE open `.srt` and export as "Plain text" and take `.text` file, now we can test contents of both files in https://wordcounter.net
Word count should match, otherwise - Houston we have a problem. 😆

**Reply by Purfview** [reply]
Btw, I accidentally noticed wrong line split in your file at 19 sub, the split at `Dr.`. 
I made ignore list from SE lists, it will be in the next release:

>  ignore_list  = ['Dr.', 'Mr.', 'Mrs.', 'Ms.', 'Frk.', 'Hr.', 'Sra.', 'Srta.', 'Dra.', 'Prof.', 'Mtro.', 'Mtra.', 'Lic.', 'Arq.', 'Ing.', 'Cnel.', 'Tte.', 'Gral.', 'dipl.', 'bl.', 'gosp.', 'ing.', 'mr.', 'prof.', 'sv.', 'vlč.', 'dr.', 'g.', 'DR.', 'DRA.', 'eng.', 'Eng.', 'exmo.', 'Exmo.', 'gov.', 'Gov.', 'jr.', 'Jr.', 'mna.', 'Mna.', 'mno.', 'Mno.', 'mrs.', 'P.', 'pe.', 'Pe.', 'srs.', 'Srs.', 'St.', 'vol.', 'vols.', 'sr.', 'Sr.']

**Comment by 1001ruchka** [comment]
Please clarify the meaning of the parameters:
--standard
--max_comma
--max_gap
unfortunately I could not find a description

**Reply by Purfview** [reply]
`--help`

**Comment by despairTK** [comment]
In version r160.10, a sentence segmentation error occurs when the --sentence parameter is enabled. as the picture shows:
![QQ截图20231203235548](https://github.com/Purfview/whisper-standalone-win/assets/111871110/5706e2de-a9ea-4958-9ede-1eae09735402)
![QQ截图20231203235648](https://github.com/Purfview/whisper-standalone-win/assets/111871110/677d39a6-cd93-4216-b55b-768b08b0b640)

Here are all the files output.
[test.zip](https://github.com/Purfview/whisper-standalone-win/files/13539512/test.zip)

My transcription settings：--model large-v2  --beam_size 10 --best_of 10 --sentence --compute_type=float16 --max_comma 1000

**Reply by Purfview** [reply]
Not error, there is over 3.0 seconds gap between `There` and `you go.`. 
`max_gap` default is 3.0.

Why there are such big gaps between words? Are you doing a translation here?

`I feel` has 10 seconds gap. 
`All right` has 80 seconds gap.  😄

**Reply by despairTK** [reply]
The problem is that the original normal sentence is strangely split after using the --sentence parameter.

**Reply by Purfview** [reply]
> The problem is that the original normal sentence is strangely split after using the --sentence parameter.

It's split correctly according to the settings.

It's not "strangely split", timestamps are "strange" in your transcription:

![Screenshot_2](https://github.com/Purfview/whisper-standalone-win/assets/69023953/a00b09f2-d9f9-4043-ae45-1e404258ffc4)

**Reply by despairTK** [reply]
terribly sorry. . . I didn't check carefully. It's a waste of your precious time. . . You can delete this feedback conversation as no real feedback was provided.

**Reply by Purfview** [reply]
Nah, let it be. Maybe I'll clean up the images.

Maybe it's not clear but:
`--sentence` activates all those new `max_....` parameters.

**Comment by Azzhoe** [comment]
what does the --standard_asia do?

**Reply by Purfview** [reply]
`--help`

**Comment by oep42** [comment]
**Suggestion 1:** Make the `--standard` parameter versatile, by enabling it to pass numeric values to its "built-in" parameters, like this:

- Without a numeric value, the current hard-coded values of its "built-in" parameters are being used, so these four parameters are activated with these values: `--sentence`, `--max_line_width 42`, `--max_line_count 2`, `--max_comma_cent 70`.

- With a numeric value such as this `--standard 38,3,50`, these four parameters are activated with these values: `--sentence`, `--max_line_width 38`, `--max_line_count 3`, `--max_comma_cent 50`.

**Suggestion 2:** Make explicit in the `--help` information what the `--standard` parameter exactly does, by explicitly mentioning all four parameters that are activated by `--standard`. In the current description, it is not obvious to those unfamiliar with the subject, that the parameters `--max_line_width` and `--max_line_count` are meant.

**Reply by Purfview** [reply]
>Suggestion 2:

Done.

>Suggestion 1

Nah, that would be a bit weird overengineering.
